File: frontend/src/repositories/unitRepository.ts

import { db } from "../db/database";
import type { Unit, UnitListItem } from "../models/domain";
import type { Repository } from "./repository";
import { applicationClock } from "../services/applicationClockService";
import { apiRequest } from "./apiClient";
import { repositoryConfiguration } from "./repositoryConfiguration";

export interface UnitRepository extends Repository<Unit> {
  getListItems(): Promise<UnitListItem[]>;
}

export class DexieUnitRepository implements UnitRepository {
  getAll(): Promise<Unit[]> {
    return db.units.toArray();
  }

  getById(id: number): Promise<Unit | undefined> {
    return db.units.get(id);
  }

  async add(entity: Omit<Unit, "id">): Promise<number> {
    return Number(await db.units.add(entity));
  }

  async update(
    id: number,
    changes: Partial<Omit<Unit, "id">>,
  ): Promise<void> {
    await db.units.update(id, changes);
  }

  async delete(id: number): Promise<void> {
    await db.units.delete(id);
  }

  async getListItems(): Promise<UnitListItem[]> {
    const today = applicationClock.today();
    const [units, buildings, locations, leases, charges] = await Promise.all([
      db.units.toArray(),
      db.buildings.toArray(),
      db.locations.toArray(),
      db.leases.toArray(),
      db.recurringCharges.toArray(),
    ]);

    const buildingMap = new Map(buildings.map((item) => [item.id, item]));
    const locationMap = new Map(locations.map((item) => [item.id, item]));

    return units.map((unit) => {
      const building = buildingMap.get(unit.buildingId);
      const location = building
        ? locationMap.get(building.locationId)
        : undefined;

      const applicableLease = leases
        .filter((lease) => {
          if (lease.unitId !== unit.id) return false;
          if (lease.status === "Expired" || lease.status === "Terminated") {
            return false;
          }

          const hasStarted = lease.startDate <= today;
          const hasNotEnded =
            lease.termType === "Month-to-Month" ||
            !lease.endDate ||
            lease.endDate >= today;

          return hasStarted && hasNotEnded;
        })
        .sort((left, right) =>
          right.startDate.localeCompare(left.startDate),
        )[0];

      const apartmentRent = applicableLease
        ? charges.find(
            (charge) =>
              charge.leaseId === applicableLease.id &&
              charge.chargeType === "Apartment Rent" &&
              charge.frequency === "Monthly" &&
              charge.startDate <= today &&
              (!charge.endDate || charge.endDate >= today),
          )?.amount
        : undefined;

      return {
        ...unit,
        street: location?.name ?? "Unknown",
        civicAddress: building?.civicAddress ?? "Unknown",
        effectiveRent: apartmentRent ?? unit.monthlyRent,
        rentSource:
          apartmentRent !== undefined ? "Active Lease" : "Market Rent",
      };
    });
  }
}

export class ApiUnitRepository implements UnitRepository {
  getAll(): Promise<Unit[]> {
    return apiRequest<UnitListItem[]>("/api/v1/units");
  }

  getById(id: number): Promise<Unit | undefined> {
    return apiRequest<Unit>(`/api/v1/units/${id}`);
  }

  async add(entity: Omit<Unit, "id">): Promise<number> {
    const created = await apiRequest<Unit>("/api/v1/units", {
      method: "POST",
      body: JSON.stringify(entity),
    });
    if (created.id === undefined) throw new Error("The backend did not return a unit identifier.");
    return created.id;
  }

  async update(id: number, changes: Partial<Omit<Unit, "id">>): Promise<void> {
    const current = await this.getById(id);
    if (!current) throw new Error("Unit not found.");
    await apiRequest<Unit>(`/api/v1/units/${id}`, {
      method: "PUT",
      body: JSON.stringify({ ...current, ...changes }),
    });
  }

  delete(id: number): Promise<void> {
    return apiRequest<void>(`/api/v1/units/${id}`, { method: "DELETE" });
  }

  getListItems(): Promise<UnitListItem[]> {
    return apiRequest<UnitListItem[]>("/api/v1/units");
  }
}

export const unitRepository: UnitRepository =
  repositoryConfiguration.units === "api"
    ? new ApiUnitRepository()
    : new DexieUnitRepository();

File: backend/property_manager/repositories/unit_repository.py
from __future__ import annotations

from decimal import Decimal
from typing import Any

from ..database import read_connection


class UnitRepository:
    """Persistence operations for units and their list-page projections."""

    def list_units(self) -> list[dict[str, Any]]:
        with read_connection() as connection:
            cursor = connection.cursor(dictionary=True)
            cursor.execute(
                """
                SELECT u.id,
                       u.building_id,
                       u.apartment_number,
                       u.bedrooms,
                       u.bathrooms,
                       u.monthly_rent,
                       u.status,
                       u.active,
                       b.civic_address,
                       l.name AS street
                  FROM units AS u
                  JOIN buildings AS b ON b.id = u.building_id
                  JOIN locations AS l ON l.id = b.location_id
                 ORDER BY l.name, b.civic_address, u.apartment_number, u.id
                """
            )
            units = list(cursor.fetchall())

            cursor.execute(
                """
                SELECT le.unit_id, rc.amount
                  FROM leases AS le
                  JOIN recurring_charges AS rc
                    ON rc.lease_id = le.id
                   AND rc.charge_type = 'Apartment Rent'
                   AND rc.frequency = 'Monthly'
                   AND rc.start_date <= CURRENT_DATE()
                   AND (rc.end_date IS NULL OR rc.end_date >= CURRENT_DATE())
                 WHERE le.status NOT IN ('Expired', 'Terminated')
                   AND le.start_date <= CURRENT_DATE()
                   AND (
                       le.term_type = 'Month-to-Month'
                       OR le.end_date >= CURRENT_DATE()
                   )
                 ORDER BY le.unit_id, le.start_date DESC, rc.start_date DESC, rc.id DESC
                """
            )
            active_rents: dict[int, Decimal] = {}
            for row in cursor.fetchall():
                active_rents.setdefault(int(row["unit_id"]), row["amount"])
            cursor.close()

        for unit in units:
            active_rent = active_rents.get(int(unit["id"]))
            unit["effective_rent"] = (
                active_rent if active_rent is not None else unit["monthly_rent"]
            )
            unit["rent_source"] = (
                "Active Lease" if active_rent is not None else "Market Rent"
            )
        return units

    def get_by_id(self, unit_id: int) -> dict[str, Any] | None:
        with read_connection() as connection:
            cursor = connection.cursor(dictionary=True)
            cursor.execute(
                """
                SELECT id, building_id, apartment_number, bedrooms, bathrooms,
                       monthly_rent, status, active
                  FROM units
                 WHERE id = ?
                """,
                (unit_id,),
            )
            row = cursor.fetchone()
            cursor.close()
        return row

    @staticmethod
    def find_duplicate(
        connection: Any,
        building_id: int,
        apartment_number: str,
        excluding_id: int | None = None,
    ) -> int | None:
        cursor = connection.cursor()
        sql = """
            SELECT id FROM units
             WHERE building_id = ? AND LOWER(apartment_number) = LOWER(?)
        """
        parameters: list[Any] = [building_id, apartment_number]
        if excluding_id is not None:
            sql += " AND id <> ?"
            parameters.append(excluding_id)
        cursor.execute(sql, parameters)
        row = cursor.fetchone()
        cursor.close()
        return int(row[0]) if row else None

    @staticmethod
    def building_exists(connection: Any, building_id: int) -> bool:
        cursor = connection.cursor()
        cursor.execute("SELECT 1 FROM buildings WHERE id = ? LIMIT 1", (building_id,))
        exists = cursor.fetchone() is not None
        cursor.close()
        return exists

    @staticmethod
    def exists(connection: Any, unit_id: int) -> bool:
        cursor = connection.cursor()
        cursor.execute("SELECT 1 FROM units WHERE id = ? LIMIT 1", (unit_id,))
        exists = cursor.fetchone() is not None
        cursor.close()
        return exists

    @staticmethod
    def next_id(connection: Any) -> int:
        cursor = connection.cursor()
        cursor.execute("SELECT id FROM units ORDER BY id DESC LIMIT 1 FOR UPDATE")
        row = cursor.fetchone()
        cursor.close()
        return (int(row[0]) + 1) if row else 1

    @staticmethod
    def insert(connection: Any, unit_id: int, values: dict[str, Any]) -> None:
        cursor = connection.cursor()
        cursor.execute(
            """
            INSERT INTO units (
                id, building_id, apartment_number, bedrooms, bathrooms,
                monthly_rent, status, active
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                unit_id,
                values["building_id"],
                values["apartment_number"],
                values["bedrooms"],
                values["bathrooms"],
                values["monthly_rent"],
                values["status"],
                values["active"],
            ),
        )
        cursor.close()

    @staticmethod
    def update(connection: Any, unit_id: int, values: dict[str, Any]) -> None:
        cursor = connection.cursor()
        cursor.execute(
            """
            UPDATE units
               SET building_id = ?, apartment_number = ?, bedrooms = ?,
                   bathrooms = ?, monthly_rent = ?, status = ?, active = ?
             WHERE id = ?
            """,
            (
                values["building_id"],
                values["apartment_number"],
                values["bedrooms"],
                values["bathrooms"],
                values["monthly_rent"],
                values["status"],
                values["active"],
                unit_id,
            ),
        )
        cursor.close()

    @staticmethod
    def has_lease_history(connection: Any, unit_id: int) -> bool:
        cursor = connection.cursor()
        cursor.execute("SELECT 1 FROM leases WHERE unit_id = ? LIMIT 1", (unit_id,))
        exists = cursor.fetchone() is not None
        cursor.close()
        return exists

    @staticmethod
    def delete(connection: Any, unit_id: int) -> bool:
        cursor = connection.cursor()
        cursor.execute("DELETE FROM units WHERE id = ?", (unit_id,))
        deleted = cursor.rowcount > 0
        cursor.close()
        return deleted
